l5kit/l5kit/environment/envs/l5_env.py

```diff
@@ -0,0 +1,200 @@
+            action = convert_to_dict(action, self.future_num_frames)
+            self.ego_output_dict = action
+
+            if not self.open_loop:
+                # Update raster according to predicted ego since closed loop training
+                self.simulator.update_ego(self.sim_dataset, next_frame_index, self.ego_input_dict, self.ego_output_dict)
+
+            ego_frame_in_out = self.simulator.get_ego_in_out(self.ego_input_dict, self.ego_output_dict,
+                                                             self.simulator.keys_to_exclude)
+            for scene_idx in self.scene_indices:
+                self.ego_ins_outs[scene_idx].append(ego_frame_in_out[scene_idx])
+
+        # reward calculation
+        if self.open_loop:
+            reward = self.reward.get_open_loop_reward(self.ego_output_dict, self.ego_input_dict)
+        else:
+            reward = self.reward.get_close_loop_reward(self.frame_index, self.scene_indices, self.sim_dataset,
+                                                       self.ego_ins_outs, self.agents_ins_outs)
+
+        # done is True when episode ends or error gets too high (optional)
+        done = episode_over or self.check_done_status()
+
+        # Optionally we can pass additional info
+        # We are using "info" to output simulated outputs
+        info = {}
+        if done:
+            info = {"info": self.get_simulated_outputs()}
+
+        # Prepare next obs
+        if not episode_over:
+            self.frame_index += 1
+            ego_input = self.sim_dataset.rasterise_frame_batch(self.frame_index)
+            self.ego_input_dict = default_collate_numpy(ego_input[0])
+            obs = {"image": ego_input[0]["image"]}
+        else:
+            # Dummy final obs (when episode_over)
+            ego_input = self.sim_dataset.rasterise_frame_batch(0)
+            self.ego_input_dict = default_collate_numpy(ego_input[0])
+            obs = {"image": ego_input[0]["image"]}
+
+        # return obs, reward, done, info
+        return GymStepOutput(obs, reward, done, info)
+
+    def check_done_status(self) -> bool:
+        """
+        (Optionally) End episode if the displacement error crosses a threshold
+        :return: end episode flag
+        """
+        return self.reward.stop_error > self.reward.stop_thresh
+
+    def get_simulated_outputs(self) -> List[SimulationOutputGym]:
+        """
+        Generate simulated outputs at end of episode
+        :return: List of simulated outputs
+        """
+        assert len(self.scene_indices) == 1
+        # generate simulated_outputs
+        simulated_outputs: List[SimulationOutputGym] = []
+        for scene_idx in self.scene_indices:
+            simulated_outputs.append(SimulationOutputGym(scene_idx, self.sim_dataset,
+                                                         self.ego_ins_outs, self.agents_ins_outs))
+        return simulated_outputs
+
+    def render(self) -> None:
+        pass
```

Comment:
I think it would be better to raise a `NotImplementError` error here, to be explicit about the support of this.